cpp/src/stream_compaction/distinct_reduce.cu

```diff
@@ -111,15 +112,26 @@ rmm::device_uvector<size_type> hash_reduce_by_row(
   auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
   auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
 
-  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
-  auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal);
-
-  thrust::for_each(
-    rmm::exec_policy(stream),
-    thrust::make_counting_iterator(0),
-    thrust::make_counting_iterator(num_rows),
-    reduce_by_row_fn{
-      map.get_device_view(), key_hasher, key_equal, keep, reduction_results.begin()});
+  auto const row_comp = cudf::experimental::row::equality::self_comparator(preprocessed_input);
+
+  auto const reduce_by_row = [&](auto const value_comp) {
```

Comment:
Along the lines of what @PointKernel was suggesting -- one alternative I considered was making this IIFE actually allocate and return the output vector, rather than binding in `reduction_results` and returning `void`. That felt like an implicit "output parameter" in the IIFE. If the IIFE were a real function, we'd avoid the output parameter and return the `rmm::device_uvector` directly from the function. I don't have strong feelings on this, so feel free to keep it as-is.

Reply:
If we did that, however, almost the entire body of `hash_reduce_by_row` becomes an IIFE and the part at the end is just a `nans_equal` dispatcher. 😛